src/Tools/dotnet-stack/Symbolicate.cs

```diff
@@ -200,54 +174,109 @@ private static string TrySymbolicateLine(string line)
-                            if (point.Offset > Convert.ToInt64(stInfo.Offset, 16))
-                                break;
-
-                            if (point.StartLine != SequencePoint.HiddenLine)
-                                bestPointSoFar = point;
+                            return reader;
                         }
+                    }
+                    /*else if (filePath.Contains(".pdb"))
+                    {
+                        provider = MetadataReaderProvider.FromPortablePdbStream(stream);
+                    }*/
+                    else
+                    {
+                        return reader;
+                    }
+                }
+                return provider?.GetMetadataReader();
+            }
+            catch
+            {
+                return reader;
+            }
+        }
 
-                        if (bestPointSoFar.HasValue)
+        private static string GetLineFromMetadata(MetadataReader reader, string line, StackTraceInfo stInfo)
+        {
+            try
+            {
+                if (reader != null)
+                {
+                    Handle handle = MetadataTokens.Handle(Convert.ToInt32(stInfo.Token, 16));
+                    if (handle.Kind == HandleKind.MethodDefinition)
+                    {
+                        MethodDebugInformationHandle methodDebugHandle = ((MethodDefinitionHandle)handle).ToDebugInformationHandle();
+                        MethodDebugInformation methodInfo = reader.GetMethodDebugInformation(methodDebugHandle);
+                        if (!methodInfo.SequencePointsBlob.IsNil)
                         {
-                            string sourceFile = reader.GetString(reader.GetDocument(bestPointSoFar.Value.Document).Name);
-                            int sourceLine = bestPointSoFar.Value.StartLine;
-                            return $"   at {stInfo.Type}.{stInfo.Method}({stInfo.Param}) in {sourceFile}:line {sourceLine}";
+                            SequencePointCollection sequencePoints = methodInfo.GetSequencePoints();
+                            SequencePoint? bestPointSoFar = null;
+                            foreach (SequencePoint point in sequencePoints)
+                            {
+                                if (point.Offset > Convert.ToInt64(stInfo.Offset, 16))
+                                    break;
+
+                                if (point.StartLine != SequencePoint.HiddenLine)
+                                    bestPointSoFar = point;
+                            }
+
+                            if (bestPointSoFar.HasValue)
+                            {
+                                string sourceFile = reader.GetString(reader.GetDocument(bestPointSoFar.Value.Document).Name);
+                                int sourceLine = bestPointSoFar.Value.StartLine;
+                                return $"   at {stInfo.Type}.{stInfo.Method}({stInfo.Param}) in {sourceFile}:line {sourceLine}";
```

Comment:
Theoretically, the input line could have any number or type of whitespace (or other characters) at the beginning of the line. For example, the input text may be from a log that has timestamps on every line. We could potentially be throwing out data by rewriting the line like this.

If possible, we should try to rewrite the line _in situ_. You could reuse the regex you already have with the [`Regex.Replace(String, String, MatchEvaluator, RegexOptions)`](https://docs.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regex.replace?view=net-5.0#System_Text_RegularExpressions_Regex_Replace_System_String_System_String_System_Text_RegularExpressions_MatchEvaluator_System_Text_RegularExpressions_RegexOptions_) API. You would write a `MatchEvaluator` method that receives each match in the regex and replaces it with the appropriate part of the `StackTraceInfo`. You may also need to add a capture group for the `: token` part so you can replace it with `: line`.